Requests are sent before HttpClientFilter finishes

### Expected Behavior

The attached test case should pass in Micronaut >= 4.8.0

### Actual Behaviour

The attached test case does not pass in >= 4.8.0.  It does pass in <= 4.7.6

The test creates a span, sets a baggage item, calls a declarative client which calls a controller which checks for said baggage item.  If the baggage item exists, the controller returns `foo` which is expected; otherwise, it returns `bar` which causes the test to fail.  The expectation is that `OpenTracingClientFilter` injects the span context into the headers before sending the request

The failure is intermittent, so I created a loop of 10k, but it usually fails around 100 iterations for me if there is 1 test case

But, it is orders of magnitude more noticeable when there are multiple test cases.  It generally reproduces on the first or second iteration within 10 test cases

When I debugged this in our product, I found that the controller was processing the request at the same time `OpenTracingClientFilter` was still injecting the baggage into the headers, so clearly there is a race condition.  It might be an issue with `OpenTracingClientFilter` but I don't see anything special about it nor do I see any real changes in it between the copies used in 4.7.6 vs 4.8.0 so it might just be a general `HttpClientFilter` issue which would seem significant

### Steps To Reproduce

```
package com.example;

import io.micronaut.context.annotation.Property;
import io.micronaut.http.annotation.Controller;
import io.micronaut.http.annotation.Get;
import io.micronaut.http.client.annotation.Client;
import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import io.opentracing.Scope;
import io.opentracing.Tracer;
import jakarta.inject.Inject;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.slf4j.LoggerFactory;

import java.util.Optional;


@MicronautTest
@Property(name = "tracing.jaeger.enabled", value = "true")
@Property(name = "tracing.jaeger.sampler.probability", value = "1")
class MissingBaggageTest {
    @Inject MyClient myClient;
    @Inject Tracer tracer;

    private void testItWorks() {
        for (int i = 0; i < 10_000; ++i) {
            try (final Scope ignore = tracer.activateSpan(tracer.buildSpan("foo").start().setBaggageItem("foo", "foo"))) {
                Assertions.assertEquals("foo", myClient.getFoo(), "Missing baggage detected on iteration " + i);
            }
        }
    }

    @Test void testItWorks1() { testItWorks(); }
    @Test void testItWorks2() { testItWorks(); }
    @Test void testItWorks3() { testItWorks(); }
    @Test void testItWorks4() { testItWorks(); }
    @Test void testItWorks5() { testItWorks(); }
    @Test void testItWorks6() { testItWorks(); }
    @Test void testItWorks7() { testItWorks(); }
    @Test void testItWorks8() { testItWorks(); }


    @Controller("/foo")
    public static class MyController {
        @Inject Tracer tracer;

        @Get
        public String getFoo() {
            return Optional
                .of(tracer)
                .map(Tracer::activeSpan)
                .map(s -> s.getBaggageItem("foo"))
                .orElseGet(() -> {
                    LoggerFactory.getLogger(getClass()).error("Missing baggage!");
                    return "bar";
                });
        }
    }

    @Client("/foo")
    public interface MyClient {
        @Get
        String getFoo();
    }
}
```

When I run on >= 4.8.0, I see:
```
[ERROR]   MissingBaggageTest.testItWorks1:33->testItWorks:28 Missing baggage detected on iteration 188 ==> expected: <foo> but was: <bar>
[ERROR]   MissingBaggageTest.testItWorks2:34->testItWorks:28 Missing baggage detected on iteration 77 ==> expected: <foo> but was: <bar>
[ERROR]   MissingBaggageTest.testItWorks3:35->testItWorks:28 Missing baggage detected on iteration 159 ==> expected: <foo> but was: <bar>
[ERROR]   MissingBaggageTest.testItWorks4:36->testItWorks:28 Missing baggage detected on iteration 298 ==> expected: <foo> but was: <bar>
[ERROR]   MissingBaggageTest.testItWorks5:37->testItWorks:28 Missing baggage detected on iteration 58 ==> expected: <foo> but was: <bar>
[ERROR]   MissingBaggageTest.testItWorks6:38->testItWorks:28 Missing baggage detected on iteration 0 ==> expected: <foo> but was: <bar>
[ERROR]   MissingBaggageTest.testItWorks7:39->testItWorks:28 Missing baggage detected on iteration 1 ==> expected: <foo> but was: <bar>
[ERROR]   MissingBaggageTest.testItWorks8:40->testItWorks:28 Missing baggage detected on iteration 18 ==> expected: <foo> but was: <bar>
```

### Environment Information

```
$> mvn --version

Apache Maven 3.8.7 (b89d5959fcde851dcb1c8946a785a163f14e1e29)
Maven home: /.user/programs/sdkman/candidates/maven/3.8.7
Java version: 21.0.2, vendor: Oracle Corporation, runtime: /.user/programs/sdkman/candidates/java/21.0.2-graal
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "5.4.17-2136.312.3.4.el7uek.x86_64", arch: "amd64", family: "unix"
```

### Example Application

_No response_

### Version

4.8.0

--- a/tracing-opentracing/build.gradle
+++ b/tracing-opentracing/build.gradle
@@ -13,5 +13,6 @@ dependencies {
     testImplementation "io.reactivex.rxjava2:rxjava"
     testImplementation mnRxjava2.micronaut.rxjava2.http.client
     testImplementation mnReactor.micronaut.reactor.http.client
+    testImplementation project(":micronaut-tracing-jaeger")
     testRuntimeOnly mnSerde.micronaut.serde.jackson
 }

--- a/tracing-opentracing/src/main/java/io/micronaut/tracing/opentracing/instrument/http/OpenTracingClientFilter.java
+++ b/tracing-opentracing/src/main/java/io/micronaut/tracing/opentracing/instrument/http/OpenTracingClientFilter.java
@@ -88,8 +88,8 @@ public final class OpenTracingClientFilter extends AbstractOpenTracingFilter imp
             .plus(new OpenTracingPropagationContext(tracer, span))
             .propagate()) {
 
+            tracer.inject(span.context(), HTTP_HEADERS, new HttpHeadersTextMap(request.getHeaders()));
             return Mono.from(chain.proceed(request))
-                .doOnSubscribe(subscription -> tracer.inject(span.context(), HTTP_HEADERS, new HttpHeadersTextMap(request.getHeaders())))
                 .doOnNext(httpResponse -> setResponseTags(request, httpResponse, span))
                 .doOnError(throwable -> {
                     if (throwable instanceof HttpClientResponseException e) {

--- a/tracing-opentracing/src/test/groovy/io/micronaut/tracing/opentracing/instrument/http/OpenTracingClientFilterSpec.groovy
+++ b/tracing-opentracing/src/test/groovy/io/micronaut/tracing/opentracing/instrument/http/OpenTracingClientFilterSpec.groovy
@@ -1,0 +1,61 @@
+package io.micronaut.tracing.opentracing.instrument.http
+
+import io.jaegertracing.internal.reporters.InMemoryReporter
+import io.micronaut.context.ApplicationContext
+import io.micronaut.context.annotation.Requires
+import io.micronaut.http.HttpRequest
+import io.micronaut.http.HttpStatus
+import io.micronaut.http.annotation.Controller
+import io.micronaut.http.annotation.Get
+import io.micronaut.http.client.HttpClient
+import io.micronaut.http.exceptions.HttpStatusException
+import io.micronaut.runtime.server.EmbeddedServer
+import io.opentracing.Scope
+import io.opentracing.Tracer
+import spock.lang.Issue
+import spock.lang.Specification
+
+import java.util.concurrent.CompletableFuture
+
+class OpenTracingClientFilterSpec extends Specification {
+    @Issue('https://github.com/micronaut-projects/micronaut-tracing/issues/784')
+    def 'subscribe after proceed'() {
+        given:
+        def ctx = ApplicationContext.builder([
+                'spec.name': 'OpenTracingClientFilterSpec',
+                'tracing.jaeger.enabled': true,
+                'tracing.jaeger.sampler.probability': 1
+        ]).singletons(new InMemoryReporter()).start()
+        def server = ctx.getBean(EmbeddedServer)
+        server.start()
+        def tracer = ctx.getBean(Tracer)
+
+        def client = ctx.createBean(HttpClient, server.URI).toBlocking()
+        def futures = new ArrayList<CompletableFuture<?>>()
+        for (int thread = 0; thread < 4; thread++) {
+            futures.add(CompletableFuture.runAsync {
+                try (Scope _ = tracer.activateSpan(tracer.buildSpan("foo").start().setBaggageItem("foo", "foo"))) {
+                    for (int i = 0; i < 10000; i++) {
+                        client.retrieve("/check-baggage")
+                    }
+                }
+            })
+        }
+        CompletableFuture.allOf(futures.toArray(CompletableFuture[]::new)).get()
+
+        cleanup:
+        ctx.close()
+    }
+
+    @Controller
+    @Requires(property = "spec.name", value = "OpenTracingClientFilterSpec")
+    static class MyController {
+        @Get("/check-baggage")
+        String checkBaggage(HttpRequest<?> request) {
+            if (request.headers.get("uberctx-foo") != "foo") {
+                throw new HttpStatusException(HttpStatus.BAD_REQUEST, "tracing baggage missing")
+            }
+            return "ok"
+        }
+    }
+}

